### DISCUSSION TITLE
Debugging in VSCode

### DISCUSSION THREAD

**Discussion by david-boles** [flxzt/rnote#148] [Development]
I haven't worked on Flatpaks or used Meson before. Is it possible to debug Rnote (running as a Flatpak?) with VSCode? If you've done it, would you mind sharing your launch configuration? Alternatively, is there a different debugging setup you'd recommend? Thanks!

**Comment by david-boles** [comment]
To clarify: I can compile and run it fine through the Flatpak extension, but haven't been able to figure out debugging.

**Comment by flxzt** [comment]
I never used debugging in the flatpak, so I don't know. For debugging a native meson build you can use the `CodeLLDB` extension.

First you'd need to change these lines in `build-aux/cargo.sh` so that cargo actually builds with debugging symbols:
```bash 
    echo -e "\n--- DEVEL PROFILE ---\n"
    cargo build --manifest-path \
        "$MESON_SOURCE_ROOT"/Cargo.toml && \
        cp "$CARGO_TARGET_DIR"/debug/"$APP_BIN" "$OUTPUT"
```

You'd need to configure, compile and install the meson project as outlined in `CONTRIBUTING.md`. Be sure to configure meson with `-Dprofile=devel`.

Then you'd create a vscode `launch.json` with something like this:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "lldb",
            "request": "launch",
            "name": "launch debug build of 'rnote'",
            "args": [],
            "program": "${workspaceFolder}/target/debug/rnote"
        },
    ]
}
```

Then you should be able to debug inside vscode, set breakpoints, etc.

Hope that helps!

**Comment by david-boles** [comment]
@flxzt thanks! I was also running into some rust-analyzer version incompatibilities so it seems like native builds are definitely the way to go. 

That mostly works... Looks like I also have to install it, otherwise I get a panic:

```
thread 'main' panicked at 'Could not load gresource file: Error { domain: g-file-error-quark, code: 4, message: "Failed to open file “/usr/share/rnote/resources.gresource”: open() failed: No such file or directory" }', rnote-ui/src/app/mod.rs:74:18
```

And I think I'd also have to re-install after every build if I change the UI? It would be nice to override where it looks for that but `config.rs` seems to get set automatically when building.

**Reply by flxzt** [reply]
You could include the meson install command in the launch.json, so it does that automatically every time when you launch it. (Could also be extended for tests, etc.)

**Reply by david-boles** [reply]
It's a little annoying since I currently have to authenticate every time; just feels a bit clunky. It looks like the meson.build script has some ability to configure the directories (e.g. `prefix` and `datadir` options) but not quite enough. Would it make sense to expand on that?

**Reply by flxzt** [reply]
I am not a fan of reconfiguring the resource path, because the build config will get messy. But you don't have to have `/usr` as prefix but can choose another path that you own (e.g. `--prefix=$(pwd)/_mesoninstall` ). You won't get the system integration, but the binary should still launch fine. Doing that currently gives another error that the pen sounds could not be loaded, but I'll fix that (and it won't panic).

**Reply by david-boles** [reply]
Ha, just figured that out myself. Thanks, this seems like a good solution to me.

**Reply by david-boles** [reply]
@flxzt You work fast! I didn't notice your updates to `CONTRIBUTING.md` so I typed up some [notes](https://github.com/david-boles/rnote/commit/bd49e7b12a2518f1a1483d0a3a0648d77146f9e2) myself. I'll revise and submit a PR but a couple questions first:
- If `/_mesoninstall` is the recommended in-place install folder, should it be added to the gitignore?
- Could `build-aux/cargo.sh` do debug builds by default on the devel profile? That seems sensible to me.

**Reply by flxzt** [reply]
Ah, your notes have some additional info so please do! One thing, you should recommend `poppler-data` as dependency as well, that provides better font rendering. (And the gstreamer dep actually was outdated and is not needed anymore)

As to your questions:
- yes, definitely. Additionally I think both `/usr` and `$(pwd)/_mesoninstall` should be mentioned as suggestions in the doc, with an explanation regarding system integration and needing elevated privileges.
- I did not do that, because running the build without `--release` was actually a lot slower for me. Maybe there should be another build profile, `debug` for it.

**Reply by david-boles** [reply]
Reminder to self: also make changes to instructions for `meson install`.

**Reply by flxzt** [reply]
fyi I tweaked the dev build profile that now does provide debugging symbols while having more code optimizations

**Reply by flxzt** [reply]
and pen sounds should now load from the pkg data dir that is configured with meson, not from glib's reported system dir anymore.
